### PR TITLE
Discard timeseries data that is outside configured time window

### PR DESCRIPTION
**Pull request by mateuszmidor**
KNTK-301
Based on https://github.com/kentik/sla_dashboard_webapp/pull/61

### CODE REVIEW

**Comment by mmac-m3a** [comment]
I think we need to first merge https://github.com/kentik/sla_dashboard_webapp/pull/57 and rebase this PR, before it can be reviewed.

File: domain/model/mesh_results.py [hunk 1]
```diff
@@ -134,6 +134,16 @@ def incremental_update(self, src: ConnectionMatrix) -> None:
             self._connections[from_agent_id] = dst_row
         self.connection_timestamp_oldest, self.connection_timestamp_newest = self._get_timestamp_range()
 
+    def drop_samples_older_than(self, threshold: datetime) -> None:
+        for row in self._connections.values():
+            for conn in row.values():
+                for i, item in enumerate(conn.health):
+                    if item.timestamp < threshold:
+                        logger.debug("Dropping %d samples outside requested time window", len(conn.health) - i)
+                        conn.health = conn.health[:i]
```
Comment (mmac-m3a) [inline domain/model/mesh_results.py:143]: It is in general unsafe to modify a list while iterating it. This code works because the iteration stops after the modification. There is though no need to modify `conn.health` while iterating it. For example like this:
```
        for row in self._connections.values():
            for conn in row.values():
                 n = 0
                 while conn.health[-1].timestamp < theshold:
                     conn.health.pop()
                     n += 1
                 if n > 0:
                     logger.debug("Dropped %d samples older than %s", n, threshold.isoformat())
```
This approach is also little bit more efficient, because:
- it does not iterate over fresh items at the front,
- it does not have to rebuild the whole list on every truncation.

File: domain/cache/caching_repo_request_driven.py [hunk 2]
```diff
@@ -132,6 +133,7 @@ def _update_cache_with(self, results: MeshResults, config: MeshConfig) -> None:
             logger.debug("Incremental cache update")
             new_results = deepcopy(current_results)
             new_results.incremental_update(results)
+            self._drop_samples_outside_timewindow(new_results)
             new_config = current_config
         else:
             logger.debug("New mesh test configuration detected. Full cache update")
```
Comment (mmac-m3a) [inline domain/cache/caching_repo_request_driven.py:139]: It should be possible to preserve accumulated history for agents that remained in the test configuration. Incremental update should use the new agent config and adjust rows and columns to match new set of agents. Worst case we end-up with a completely new matrix :).

This can be implemented in a separate PR.

**Review by mmac-m3a — CHANGES_REQUESTED** [review]
Please improve the `conn.health` truncation

File: domain/cache/caching_repo_request_driven.py (hunk 2 above)
Comment (mateuszmidor) [reply domain/cache/caching_repo_request_driven.py:139]: Yes it should be possible. I can do it if you find it worthwhile

**Review by mmac-m3a — APPROVED** [review]
